### ISSUE TITLE
New -Xshareclasses layer sub-options

### ISSUE THREAD

**Issue by hangshao0** [eclipse/openj9-docs#368]
**Issue or pull request number:**
https://github.com/eclipse/openj9/pull/7213
https://github.com/eclipse/openj9/pull/7270
https://github.com/eclipse/openj9/pull/7351

**Overview:**
3 new -Xshareclasses sub-options are added. 
`-Xshareclasses:layer=<number>`              Specify the top layer number. Create a new shared cache layer if the specified layer does not exist.
`-Xshareclasses:createLayer`        Create a new shared cache layer.
`-Xshareclasses:printTopLayerStats[=option[+s]]`     Print summary of cache statistics from top layer only. Use `printTopLayerStats=help` to see available options.
By default, the current options sub-options `printStats`, `printStats=` and `printAllStats` shows statistics from all layers.

These 3 options are for users running Java inside docker containers where user wants to build shared caches into different layers of the docker image.

Layered shared caches are a preview feature in the 0.17 release, and the doc should indicate this. There is still some work to do to complete the feature, including the addition of testing. Details of the limitations will be provided in time to document for the 0.17 release. People are free to try it out and use it with the known limitations.

**Release target:**
*Eclipse OpenJ9 0.17.0*

**Applies to the following JDK versions:**
*All versions*

**Applies to the following platforms:**
*All 64-bit platforms*

**For new command line options:**

- Option name and syntax: 
`-Xshareclasses:layer=<number>`, where `number` is a integer >=0. 
`-Xshareclasses:createLayer`.
`-Xshareclasses:printTopLayerStats[=option[+s]]`. The same as `printStat` and `printStats=`, It can be used as `-Xshareclasses:printTopLayerStats` or something like `-Xshareclasses:printTopLayerStats=romclass+aot`. Available options are:

```
java -Xshareclasses:printtoplayerStats=help
Available options for -Xshareclasses:printTopLayerStats=<options>. Use '+' to specify multiple options. i.e. printTopLayerStats=aot+url

         all            Prints all the following types in the shared cache.
         classpath      Prints only classpath types in the shared cache.
         url            Prints only url types in the shared cache.
         token          Prints only token types in the shared cache.
         romclass       Prints only romclass types in the shared cache.
         rommethod      Prints only rommethod types in the shared cache.
         aot            Prints only aot types in the shared cache.
         invalidatedaot Prints only invalidated aot types in the shared cache.
         jitprofile     Prints only jitprofile types in the shared cache.
         jithint        Prints only jithint types in the shared cache.
         zipcache       Prints only zipcache types in the shared cache.
         stale          Prints all the stale entries in the shared cache.
         startuphint    Prints only startup hint types in the shared cache.
```

- Purpose:
With this change, one JVM can be start-up more than one shared caches. These caches are with the same name but under different layers. For example. if there are no existing shared cache:

1. `Java -Xshareclasses:name=Cache1 HelloWorld ` starts up a layer 0 shared cache with name 'Cache1'
`listAllCaches` will show the cache.

```
java -Xshareclasses:listAllCaches
...
Cache name              level         cache-type      feature         layer       OS shmid       OS semid       last detach time

Compatible shared caches
Cache1                  Java8 64-bit  persistent      cr              0                                         Mon Sep 23 11:41:04 2019

```
2. Then if user wants to create an another layer, it can be done:
a) Either using `createLayer `sub-option: `Java -Xshareclasses:name=Cache1,createLayer HelloWorld`, 
b) Or `layer=<number>` sub-option. As listAllCaches shows Cache1 has layer 0 only, so user can create a layer 1 by: `Java -Xshareclasses:name=Cache1,layer=1 HelloWorld`. `listAllCaches` will show the cache of all layers.
```
java -Xshareclasses:listAllCaches
...
Cache name              level         cache-type      feature         layer       OS shmid       OS semid       last detach time

Compatible shared caches
Cache1                  Java8 64-bit  persistent      cr              0                                         Mon Sep 23 11:41:04 2019
Cache1                  Java8 64-bit  persistent      cr              1                                         Mon Sep 23 11:46:25 2019
```
If there are more than one JVMs in a race condition creating a layer, `layer=<number>` ensures only one new layer is created. 

The cache of different layers will be created in the same cache directory. Subsequent runs with `-Xshareclasses:name=Cache1` will startup all the layers of Cache1 (layer 0 and layer 1 in this example). Please note that only the top layer (layer 1 in this example) is started up in read-write mode. All the lower layers (layer 0 in this example) are started up in read-only mode. No modifications are allowed to a lower layer cache, doing so will invalidate all the upper layer caches. 

User can use `java -Xshareclasses:name=Cache1,destroy` to destroy the top-layer or ` java -Xshareclasses:name=Cache1,destroyAllLayers` to destroy all layers.

- Restrictions:
N.A.

**Comment by pshipton** [comment]
What is the use case for `layer=<number>`?

**Comment by hangshao0** [comment]
> What is the use case for `layer=<number>`? 

If there are more than one JVMs in a race condition creating a layer, `layer=<number>` ensures only one new layer is created. (Added this in the description.)

**Comment by doveye** [comment]
My understanding of Docker is somewhat tenuous so I have some questions :) 
- Are we talking about when building a Docker image or when running a Docker container? 
- There are quite a few mentions of creating layers. Do you really mean that, or do you mean creating shared classes caches in existing Docker layers in a running container? 
- How do the layer numbers correspond to the Docker layers? So in the Docker example at https://docs.docker.com/storage/storagedriver/, which has 5 layers, which is layer 0, layer 1 etc?

I think it would help if we provided an example, similar to the Docker one, to explain this concept to users and help them understand when/why they would do this.

**Comment by hangshao0** [comment]
> Are we talking about when building a Docker image or when running a Docker container?

A container is a container (read-write) layer on top of one or more image (read-only) layers. Users can use the new options to create a new layer shared cache and build this shared cache into their docker images. By doing this, user can avoid writing into the per-existing cache in a lower image layer. Docker handles all write operation to a image layer by a copy-on-write mechanism, which will copy the whole shared cache from a lower image to the container layer and make writes happen in the container layer, which is undesirable as we are duplicating the shared cache here. With these options, user can create a new layer cache in the container layer and all the new data will be written into the new layer cache. The per-existing caches in image layers will be started up in read-only mode and they won't be duplicated. 

For example, we can build and include a shared cache in the OpenJ9 Docker image which is a layer 0 shared cache. Liberty is built onto of OpenJ9, so Liberty can create a layer 1 shared cache and include the layer 1 cache in the Liberty docker image. An app running on liberty can build a layer 2 cache and include it into its image. 


```
A container running a liberty app:
                                    _______________________________
Liberty app (layer 2 cache)        |_______________________________|    <---------- container layer (read write)
Liberty (with a layer 1 cache)     |_______________________________|    <--------|    
OpenJ9 (with a layer 0 cache)      |_______________________________|    <--------|    image layers (read-only)
Ubuntu 16.04                       |_______________________________|    <--------|
                              
```

JVM starts up the top layer (layer 2 here) cache in read-write mode and all the lower layers (layer 0 and 1 here) is in read-only mode. 

> There are quite a few mentions of creating layers. Do you really mean that, or do you mean creating shared classes caches in existing Docker layers in a running container?

Creating shared classes caches in existing Docker layers in a running container

> How do the layer numbers correspond to the Docker layers?

See the above example, the layer number does not correspond to the docker layers exactly.

**Comment by DanHeidinga** [comment]
This feature benefits containers but isn't limited to containers.  Users can benefit from this even in other forms of deployments

**Comment by doveye** [comment]
Thanks @hangshao0 , that helped a lot.

I've added a new section in What's New for this:
https://eclipse.github.io/openj9-docs-staging/375/version0.17/ 

And these new sections in the -Xshareclasses topic:
- https://eclipse.github.io/openj9-docs-staging/375/xshareclasses/#createlayer (unfortunately the codeblocks don't scroll here, though they do elsewhere in the docs. Could be a restriction in the MkDocs tool)
- https://eclipse.github.io/openj9-docs-staging/375/xshareclasses/#destroyalllayers
- https://eclipse.github.io/openj9-docs-staging/375/xshareclasses/#layer-experimental-64-bit-only 

Please let me know if anything needs changing, thanks.

**Comment by hangshao0** [comment]
https://eclipse.github.io/openj9-docs-staging/375/xshareclasses/#createlayer

> Because each cache builds on the cache in the layer below, rather than duplicating it, space is saved in the image.

This sentence looks unclear to me. Probably say something like:
Writing to an existing shared cache in a lower image layer will lead docker to duplicate the shared cache. Instead we write into a new cache in the container layer.  

> As part of this image, the -Xhsareclasses:name=Cache1,createLayer suboption is used

Typo: -Xhsareclasses. 

This example talks about Open Liberty. In Liberty there are actually 2 JVMs been started up, so in fact using `createLayer` will result in 2 new layers (layer 1 and layer 2) been created. To create layer 1 cache in Liberty, the correct CML option should  be `-Xshareclasses:name=Cache1,layer=1`. 

> The caches in the lower layers are read only, because modifying them would invalidate all the caches in the layers above.

Probably change it to: The cache in the top layer is started up in read-write mode and caches in lower layers are started up in read-only mode. Modifying a lower layer cache would invalidate all the caches in the layers above.

**Comment by pshipton** [comment]
> https://eclipse.github.io/openj9-docs-staging/375/version0.17/
> Two new suboptions, createLayer and layer=<number>, are available for creating layered caches in a container.

Please remove "in a container". A container is not required to create layered caches. Similarly in xshareclasses/#createlayer and xshareclasses/#layer

> https://eclipse.github.io/openj9-docs-staging/375/xshareclasses/#createlayer
> however only the cache in the container layer is writable. The caches in the lower layers are read only

however only the top layer cache is writable. The lower layer caches are read only

**Comment by doveye** [comment]
Jenkins doc stage

**Comment by doveye** [comment]
Thanks for the reviews, I've updated both topics as a result.

I left in the bit about saving space as I think it's important that we state the benefit of using a layered cache, and space saving seems to be the benefit of doing so in a Docker scenario.

I've moved the example to the layer=n suboption, since the example now talks about that. I added the bit about Open Liberty starting 2 VMs, as that is an example of why the layer=n suboption is sometimes a better choice than createLayer.

Is there a simple example of a situation where a user would want to use a layered cache outside of a container?

**Comment by hangshao0** [comment]
> Is there a simple example of a situation where a user would want to use a layered cache outside of a container?

I don't see a use case mmediately, but we won't prevent user from using layered cache outside of a container.

**Comment by pshipton** [comment]
The updated doc lgtm.

**Comment by pshipton** [comment]
> Is there a simple example of a situation where a user would want to use a layered cache outside of a container?

I don't think it needs to go into the doc, but a product could install itself and create a layer 0 cache for the installation. The product installation could be read-only, and then users can create a layer 1 cache to further optimize.

**Comment by doveye** [comment]
Thanks. Have made a couple of minor tweaks (to retarget links to `layer` instead of `createLayer`), so this one is done as soon as the PR is merged.

**Comment by doveye** [comment]
Thanks for merging the PR Peter (and checking it!) Closing this issue.

**Comment by doveye** [comment]
Actually...I believe there will still be some limitations to document?

**Comment by hangshao0** [comment]
> Actually...I believe there will still be some limitations to document?

Yes. These 2 items in the task list of https://github.com/eclipse/openj9/issues/5480 are not completed yet.

Currently -Xshareclass:printStats only shows stats of top layer cache. You need to use -Xshareclasses:layer=<lowerLayerNumber>,printStats to show stats of a low layer cache. Ideally -Xshareclass:printStats should show stats of all layer caches. (PR: https://github.com/eclipse/openj9/pull/7270/)

When iterating caches and setting SH_OSCache_Info.isCorrupt, we startup the cache and call readCache() in SH_CacheMap::startupForStats() to detect corruptions. Currently we are doing this only in layer 0 cache.

**Comment by pshipton** [comment]
Likely we will include eclipse/openj9#7270 in the 0.17 release, hopefully delivered sometime this week.

**Comment by doveye** [comment]
Ok, so the second item seems very internal. What does it mean for customers? "Corruptions are only detected in the layer 0 cache."? What happens if a different later layer cache has a corruption - does the VM stop (maybe with no message)?

**Comment by hangshao0** [comment]
> What happens if a different later layer cache has a corruption - does the VM stop (maybe with no message)?

No, The VM won't stop. Users use jvmtiSharedCacheInfo and SharedClassCacheInfo to iterate all the caches in the cache directory. Users won't be able to detect a corrupted cache (whose layer > 0) by checking `jvmtiSharedCacheInfo.isCorrupt` and `SharedClassCacheInfo.isCacheCorrupt() `

If there is a VM directly starting up a corrupted cache using -Xshareclasses:name=corruptedCache, the VM will directly detect the corruption and print error messages as usual, no matter what its layer is.

**Comment by hangshao0** [comment]
Once https://github.com/eclipse/openj9/pull/7270 is merged. The example of `printStats `here needs to be updated;

https://www.ibm.com/support/knowledgecenter/en/SSYKE2_7.0.0/com.ibm.java.lnx.70.doc/diag/tools/shcpd_out_printstats.html

`Startup hint bytes` is missing in the example. With https://github.com/eclipse/openj9/pull/7270, the order of some entries is changed.

**Comment by hangshao0** [comment]
> https://www.ibm.com/support/knowledgecenter/en/SSYKE2_7.0.0/com.ibm.java.lnx.70.doc/diag/tools/shcpd_out_printstats.html

The link should be: https://www.ibm.com/support/knowledgecenter/en/SSYKE2_8.0.0/com.ibm.java.vm.80.doc/docs/shrc_pd_out_printstats.html

**Comment by pshipton** [comment]
@hangshao0 pls provide details of the new printStats option for https://github.com/eclipse/openj9/pull/7270

**Comment by hangshao0** [comment]
> pls provide details of the new printTopLayerStats options for eclipse/openj9#7270

Done.

**Comment by doveye** [comment]
- Added the "isCorrupt" restriction to https://eclipse.github.io/openj9-docs-staging/385/xshareclasses/#layer 
- Updated the What's New with the new printAllStatsOption - https://eclipse.github.io/openj9-docs-staging/385/version0.17/#new-shared-classes-cache-suboptions-for-layered-caches 
- As the existinng printStats options needed updating, and they were still in the IBM SDK docs, I moved them to OpenJ9 and rewrote as much as possible. The three options are now in a single topic, please review - https://eclipse.github.io/openj9-docs-staging/385/shrc_diag_util/ 

More questions:
- Can you please post into this issue some updated output for printStats and printTopLayerStats as there appear to have been changes since output was posted in https://github.com/eclipse/openj9/pull/7270#issuecomment-536764368, and I can't piece together what it should be from the comments.
- Same for Javacore, which I assume should be updated even though it's not mentioned in this issue. I've added the example from https://github.com/eclipse/openj9/pull/7270#issuecomment-536029586 to  https://www.eclipse.org/openj9/docs/dump_javadump/#shared-classes rather than replacing what was already there, since they are quite different and I assume the existing example still applies for non-layered caches. But again, it looks like it's changed since it was posted in 7270.
- An earlier comment in this issue said "Ideally -Xshareclass:printStats should show stats of all layer caches.", however judging by the example output some stats are shown for the top layer cache only, and some are shown for all layers combined. So there's actually no view of each layer cache individually. I changed the docs to match - is this correct?
- There's no mention in 7270 about the printAllStats output - how do layered caches affect that? I've assumed info is shown for all cache layers so I've mentioned that in the docs.

**Comment by hangshao0** [comment]
> Can you please post into this issue some updated output for printStats and printTopLayerStats as there appear to have been changes since output was posted in eclipse/openj9#7270 (comment), and I can't piece together what it should be from the comments.

On cache with two layers:

```
java -Xshareclasses:name=Cache1,printStats

Current statistics for top layer of cache "Cache1":

Cache created with:
        -Xnolinenumbers                      = false
        BCI Enabled                          = true
        Restrict Classpaths                  = false
        Feature                              = cr


base address                         = 0x00007FCAB2766000
end address                          = 0x00007FCAB374A000
allocation pointer                   = 0x00007FCAB2766000

cache layer                          = 1
cache size                           = 16776608
softmx bytes                         = 16776608
free bytes                           = 15299372
Reserved space for AOT bytes         = -1
Maximum space for AOT bytes          = -1
Reserved space for JIT data bytes    = -1
Maximum space for JIT data bytes     = -1
Class debug area size                = 1331200
Class debug area used bytes          = 0
Class debug area % used              = 0%

Cache is 8% full

Cache is accessible to current user = true
---------------------------------------------------------

Current statistics for all layers of cache "Cache1":

ROMClass bytes                       = 1328040
AOT bytes                            = 128924
JIT data bytes                       = 812
Zip cache bytes                      = 1133704
Startup hint bytes                   = 0
Data bytes                           = 114080

# ROMClasses                         = 452
# AOT Methods                        = 20
# Classpaths                         = 1
# URLs                               = 0
# Tokens                             = 0
# Zip caches                         = 21
# Startup hints                      = 0
# Stale classes                      = 0
% Stale classes                      = 0%

java -Xshareclasses:name=Cache1,printToplayerStats

Current statistics for cache "Cache1":

Cache created with:
        -Xnolinenumbers                      = false
        BCI Enabled                          = true
        Restrict Classpaths                  = false
        Feature                              = cr


base address                         = 0x00007F234C054000
end address                          = 0x00007F234D038000
allocation pointer                   = 0x00007F234C054000

cache size                           = 16776608
softmx bytes                         = 16776608
free bytes                           = 15299372
Reserved space for AOT bytes         = -1
Maximum space for AOT bytes          = -1
Reserved space for JIT data bytes    = -1
Maximum space for JIT data bytes     = -1
Metadata bytes                       = 792
Metadata % used                      = 0%
Class debug area size                = 1331200
Class debug area used bytes          = 0
Class debug area % used              = 0%

ROMClass bytes                       = 0
AOT bytes                            = 30520
JIT data bytes                       = 644
Zip cache bytes                      = 0
Startup hint bytes                   = 0
Data bytes                           = 114080

# ROMClasses                         = 0
# AOT Methods                        = 18
# Classpaths                         = 0
# URLs                               = 0
# Tokens                             = 0
# Zip caches                         = 0
# Startup hints                      = 0
# Stale classes                      = 0
% Stale classes                      = 0%


Cache is 8% full

Cache is accessible to current user = true
```

On a cache with only one layer (layer is 0):

```
java -Xshareclasses:name=Cache1,printStats

Current statistics for cache "Cache1":

Cache created with:
        -Xnolinenumbers                      = false
        BCI Enabled                          = true
        Restrict Classpaths                  = false
        Feature                              = cr

Cache contains only classes with line numbers

base address                         = 0x00007F60B807A000
end address                          = 0x00007F60B905E000
allocation pointer                   = 0x00007F60B81BE3A8

cache size                           = 16776608
softmx bytes                         = 16776608
free bytes                           = 12740572
Reserved space for AOT bytes         = -1
Maximum space for AOT bytes          = -1
Reserved space for JIT data bytes    = -1
Maximum space for JIT data bytes     = -1
Metadata bytes                       = 30440
Metadata % used                      = 0%
Class debug area size                = 1331200
Class debug area used bytes          = 189358
Class debug area % used              = 14%

ROMClass bytes                       = 1328040
AOT bytes                            = 98404
JIT data bytes                       = 168
Zip cache bytes                      = 1133704
Startup hint bytes                   = 0
Data bytes                           = 114080

# ROMClasses                         = 452
# AOT Methods                        = 2
# Classpaths                         = 1
# URLs                               = 0
# Tokens                             = 0
# Zip caches                         = 21
# Startup hints                      = 0
# Stale classes                      = 0
% Stale classes                      = 0%


Cache is 24% full

Cache is accessible to current user = true
```

So for cache with only one layer, in the output of printStats, the order of some entries has been changed compared to the old format here: https://www.ibm.com/support/knowledgecenter/en/SSYKE2_8.0.0/com.ibm.java.vm.80.doc/docs/shrc_pd_out_printstats.html. For example, entries `ROMClass bytes` and `AOT bytes` are been move down.

**Comment by hangshao0** [comment]
Javacore with one layer cache:

```
NULL           ------------------------------------------------------------------------
0SECTION       SHARED CLASSES subcomponent dump routine
NULL           ========================================
NULL
1SCLTEXTCRTW   Cache Created With
NULL           ------------------
NULL
2SCLTEXTXNL        -Xnolinenumbers       = false
2SCLTEXTBCI        BCI Enabled           = true
2SCLTEXTBCI        Restrict Classpaths   = false
NULL
1SCLTEXTCSUM   Cache Summary
NULL           ------------------
NULL
2SCLTEXTNLC        No line number content                    = false
2SCLTEXTLNC        Line number content                       = true
NULL
2SCLTEXTRCS        ROMClass start address                    = 0x00007FD2A449B000
2SCLTEXTRCE        ROMClass end address                      = 0x00007FD2A45DF0C0
2SCLTEXTMSA        Metadata start address                    = 0x00007FD2A521B8EC
2SCLTEXTCEA        Cache end address                         = 0x00007FD2A547F000
2SCLTEXTRTF        Runtime flags                             = 0x80102001ECA602BB
2SCLTEXTCGN        Cache generation                          = 41
NULL
2SCLTEXTCSZ        Cache size                                = 16776608
2SCLTEXTSMB        Softmx bytes                              = 16776608
2SCLTEXTFRB        Free bytes                                = 12830764
2SCLTEXTRCB        ROMClass bytes                            = 1327296
2SCLTEXTAOB        AOT code bytes                            = 8832
2SCLTEXTADB        AOT data bytes                            = 272
2SCLTEXTAHB        AOT class hierarchy bytes                 = 1128
2SCLTEXTATB        AOT thunk bytes                           = 0
2SCLTEXTARB        Reserved space for AOT bytes              = -1
2SCLTEXTAMB        Maximum space for AOT bytes               = -1
2SCLTEXTJHB        JIT hint bytes                            = 212
2SCLTEXTJPB        JIT profile bytes                         = 0
2SCLTEXTJRB        Reserved space for JIT data bytes         = -1
2SCLTEXTJMB        Maximum space for JIT data bytes          = -1
2SCLTEXTNOB        Java Object bytes                         = 0
2SCLTEXTZCB        Zip cache bytes                           = 1134016
2SCLTEXTSHB        Startup hint bytes                        = 0
2SCLTEXTRWB        ReadWrite bytes                           = 114080
2SCLTEXTJCB        JCL data bytes                            = 0
2SCLTEXTBDA        Byte data bytes                           = 0
2SCLTEXTMDA        Metadata bytes                            = 28808
2SCLTEXTDAS        Class debug area size                     = 1331200
2SCLTEXTDAU        Class debug area % used                   = 14%
2SCLTEXTDAN        Class LineNumberTable bytes               = 125830
2SCLTEXTDAV        Class LocalVariableTable bytes            = 63496
NULL
2SCLTEXTNRC        Number ROMClasses                         = 452
2SCLTEXTNAM        Number AOT Methods                        = 5
2SCLTEXTNAD        Number AOT Data Entries                   = 1
2SCLTEXTNAH        Number AOT Class Hierarchy                = 19
2SCLTEXTNAT        Number AOT Thunks                         = 0
2SCLTEXTNJH        Number JIT Hints                          = 5
2SCLTEXTNJP        Number JIT Profiles                       = 0
2SCLTEXTNCP        Number Classpaths                         = 1
2SCLTEXTNUR        Number URLs                               = 0
2SCLTEXTNTK        Number Tokens                             = 0
2SCLTEXTNOJ        Number Java Objects                       = 0
2SCLTEXTNZC        Number Zip Caches                         = 21
2SCLTEXTNSH        Number Startup Hint Entries               = 0
2SCLTEXTNJC        Number JCL Entries                        = 0
2SCLTEXTNST        Number Stale classes                      = 0
2SCLTEXTPST        Percent Stale classes                     = 0%
NULL
2SCLTEXTCPF        Cache is 23% full
NULL
1SCLTEXTCMST   Cache Memory Status
NULL           ------------------
1SCLTEXTCNTD       Cache Name                    Feature                  Memory type              Cache path
NULL
2SCLTEXTCMDT       Cache1                        CR                       Memory mapped file       /tmp/javasharedresources/C290M4F1A64P_Cache1_G41L00
NULL
1SCLTEXTCMST   Cache Lock Status
NULL           ------------------
1SCLTEXTCNTD       Lock Name                     Lock type                TID owning lock
NULL
2SCLTEXTCWRL       Cache write lock              File lock                Unowned
2SCLTEXTCRWL       Cache read/write lock         File lock                Unowned
NULL
NULL
NULL           ------------------------------------------------------------------------
```

Javacore with 2 layer cache:
```
NULL           ------------------------------------------------------------------------
0SECTION       SHARED CLASSES subcomponent dump routine
NULL           ========================================
NULL
1SCLTEXTCSTL   Cache Statistics for Top Layer
NULL
1SCLTEXTCRTW       Cache Created With
NULL               ------------------
NULL
2SCLTEXTXNL            -Xnolinenumbers       = false
2SCLTEXTBCI            BCI Enabled           = true
2SCLTEXTBCI            Restrict Classpaths   = false
NULL
1SCLTEXTCSUM       Cache Summary
NULL               ------------------
NULL
2SCLTEXTNLC            No line number content                    = false
2SCLTEXTLNC            Line number content                       = true
NULL
2SCLTEXTRCS            ROMClass start address                    = 0x00007F401B1BA000
2SCLTEXTRCE            ROMClass end address                      = 0x00007F401B1DD010
2SCLTEXTMSA            Metadata start address                    = 0x00007F401C050CC0
2SCLTEXTCEA            Cache end address                         = 0x00007F401C19E000
2SCLTEXTRTF            Runtime flags                             = 0x80102001ECA602BB
2SCLTEXTCGN            Cache generation                          = 41
2SCLTEXTCLY            Cache layer                               = 1
NULL
2SCLTEXTCSZ            Cache size                                = 16776608
2SCLTEXTSMB            Softmx bytes                              = 16776608
2SCLTEXTFRB            Free bytes                                = 15154352
2SCLTEXTARB            Reserved space for AOT bytes              = -1
2SCLTEXTAMB            Maximum space for AOT bytes               = -1
2SCLTEXTJRB            Reserved space for JIT data bytes         = -1
2SCLTEXTJMB            Maximum space for JIT data bytes          = -1
2SCLTEXTRWB            ReadWrite bytes                           = 1140800
2SCLTEXTDAS            Class debug area size                     = 1331200
2SCLTEXTDAU            Class debug area % used                   = 0%
2SCLTEXTDAN            Class LineNumberTable bytes               = 11748
2SCLTEXTDAV            Class LocalVariableTable bytes            = 410
NULL
2SCLTEXTCPF            Cache is 9% full
NULL
1SCLTEXTCMST       Cache Memory Status
NULL               ------------------
1SCLTEXTCNTD           Cache Name                    Feature                  Memory type              Cache path
NULL
2SCLTEXTCMDT           Cache1                        CR                       Memory mapped file       /tmp/javasharedresources/C290M4F1A64P_Cache1_G41L01
NULL
1SCLTEXTCMST       Cache Lock Status
NULL               ------------------
1SCLTEXTCNTD           Lock Name                     Lock type                TID owning lock
NULL
2SCLTEXTCWRL           Cache write lock              File lock                Unowned
2SCLTEXTCRWL           Cache read/write lock         File lock                Unowned
NULL
1SCLTEXTCSAL   Cache Statistics for All Layers
NULL
2SCLTEXTRCB            ROMClass bytes                            = 1471416
2SCLTEXTAOB            AOT code bytes                            = 152848
2SCLTEXTADB            AOT data bytes                            = 272
2SCLTEXTAHB            AOT class hierarchy bytes                 = 4128
2SCLTEXTATB            AOT thunk bytes                           = 668
2SCLTEXTJHB            JIT hint bytes                            = 456
2SCLTEXTJPB            JIT profile bytes                         = 1400
2SCLTEXTNOB            Java Object bytes                         = 0
2SCLTEXTZCB            Zip cache bytes                           = 1133704
2SCLTEXTSHB            Startup hint bytes                        = 120
2SCLTEXTJCB            JCL data bytes                            = 0
2SCLTEXTBDA            Byte data bytes                           = 0
NULL
2SCLTEXTNRC            Number ROMClasses                         = 503
2SCLTEXTNAM            Number AOT Methods                        = 31
2SCLTEXTNAD            Number AOT Data Entries                   = 1
2SCLTEXTNAH            Number AOT Class Hierarchy                = 64
2SCLTEXTNAT            Number AOT Thunks                         = 12
2SCLTEXTNJH            Number JIT Hints                          = 14
2SCLTEXTNJP            Number JIT Profiles                       = 17
2SCLTEXTNCP            Number Classpaths                         = 1
2SCLTEXTNUR            Number URLs                               = 0
2SCLTEXTNTK            Number Tokens                             = 0
2SCLTEXTNOJ            Number Java Objects                       = 0
2SCLTEXTNZC            Number Zip Caches                         = 21
2SCLTEXTNSH            Number Startup Hint Entries               = 1
2SCLTEXTNJC            Number JCL Entries                        = 0
2SCLTEXTNST            Number Stale classes                      = 0
2SCLTEXTPST            Percent Stale classes                     = 0
NULL
NULL
NULL           ------------------------------------------------------------------------
```

For javacore with one layer cache, compared to old format here: https://www.eclipse.org/openj9/docs/dump_javadump/#shared-classes. The order of some entries are changed. Entries like `ROMClass bytes` and `AOT code bytes` are moved down

**Comment by hangshao0** [comment]
> An earlier comment in this issue said "Ideally -Xshareclass:printStats should show stats of all layer caches.", however judging by the example output some stats are shown for the top layer cache only, and some are shown for all layers combined. So there's actually no view of each layer cache individually. I changed the docs to match - is this correct?

It is more accurate to say in a multi-layer cache case, `-Xshareclass:printStats` shows stats for top layer and a summary info for all layers. When `-Xshareclass:printStats=` is used, for examaple `-Xshareclass:printStats=romclass` , romclasses from all layers will be shown.

> There's no mention in 7270 about the printAllStats output - how do layered caches affect that? I've assumed info is shown for all cache layers so I've mentioned that in the docs.


Yes. `printAllStats` shows info for all cache layers, so probably there is no need to document any change for `printAllStats`

**Comment by pshipton** [comment]
Pls add a description added for 
- `Startup hint bytes` - the number of bytes of data stored to describe startup hints
- `cache layer` - the layer number the cache stats are for

`Metadata bytes` is only available in the top layer cache output, or when the cache isn't layered. There should be a note.

There should be a note that `printTopLayerStats` can be used with `layer=` to view the stats of any layer.

For `printAllStats` I think this statement is confusing `For layered caches, some information is shown for the top layer cache only, and some is shown for all layers combined.` `printAllStats` shows detailed information for all layers, it's only the `printStats summary at the end where the first section is for the top layer (size, softmx, reserved/max space, debug data size), and the second section (bytes and counts) is for all layers.

**Comment by pshipton** [comment]
Discussed with Hang to add the `cache layer` line to printStats/printTopLayerStats output as well, since printTopLayerStats can be used with `layer=`.

**Comment by pshipton** [comment]
In the `Here are example entries for various types of cache data` section, we need to add startup hints.

**Comment by pshipton** [comment]
I noticed `Stale classes` shows `because of an operating system update`. This should be because of a JVM or application update.

**Comment by pshipton** [comment]
The change to always include the `cache layer` in the printStats family output has been merged, for 0.17 as well. Hang has updated the examples output above.

**Comment by SueChaplain** [comment]
>In the Here are example entries for various types of cache data section, we need to add startup hints.

@pshipton We need some data to go with this. At the moment, I've put it underneath JIT data - not sure where it gets printed in the output. Please advise.

Further PR in place to handle many of the preceeding comments: https://github.com/eclipse/openj9-docs/pull/387

**Comment by pshipton** [comment]
Here is example startup hint data. I cut out some of the output, and we can trim it more if necessary.
```
1: 0x000000002237C6E0 STARTUP HINTS KEY: -Xoptionsfile=C:\openj9\jdk8u232-b07\jre\bin\compressedrefs\options.default -Xlockword:mode=default -Xjcl:jclse29 -Dcom.ibm.oti.vm.bootstrap.library.path=C:\openj9\jdk8u232-b07\jre\bin\compressedrefs;C:\openj9\jdk8u232-b07\jre\bin -Djava.home=C:\openj9\jdk8u232-b07\jre -Djava.ext.dirs=C:\openj9\jdk8u232-b07\jre\lib\ext -Duser.dir=C:\openj9\jdk8u232-b07\bin -Djava.class.path=. -Dsun.java.launcher=SUN_STANDARD Address: 0x000000002237C700 Size: 96
STARTUP HINTS DETAIL Flags: 1 DATA1: 1677721 DATA2: 5033165
```

**Comment by pshipton** [comment]
This comment wasn't addressed https://github.com/eclipse/openj9-docs/issues/368#issuecomment-538456664. Should I open another issue for this since it's not related to layered caches?

**Comment by SueChaplain** [comment]
>This comment wasn't addressed #368 (comment). Should I open another issue for this since it's not related to layered caches?

Fixed in https://github.com/eclipse/openj9-docs/pull/387
